File: app/src/main/java/cn/intret/app/picgo/ui/adapter/FolderListAdapterUtils.java
package cn.intret.app.picgo.ui.adapter;

import android.support.annotation.NonNull;

import com.annimon.stream.Stream;

import java.util.LinkedList;
import java.util.List;

import cn.intret.app.picgo.model.FolderModel;

public class FolderListAdapterUtils {

    @NonNull
    public static SectionedFolderListAdapter folderModelToSectionedFolderListAdapter(FolderModel model) {
        List<SectionedFolderListAdapter.Section> sections = new LinkedList<>();

        List<FolderModel.ContainerFolder> containerFolders = model.getContainerFolders();
        for (int i = 0, s = containerFolders.size(); i < s; i++) {
            sections.add(parentFolderToItem(containerFolders.get(i)));
        }

        return new SectionedFolderListAdapter(sections).setShowInFilterMode(model.isT9FilterMode());
    }

    private static SectionedFolderListAdapter.Section parentFolderToItem(FolderModel.ContainerFolder containerFolder) {
        SectionedFolderListAdapter.Section section = new SectionedFolderListAdapter.Section();

        return section.setName(containerFolder.getName())
                .setFile(containerFolder.getFile())
                .setItems(
                        Stream.of(containerFolder.getFolders())
                                .map(item -> new SectionedFolderListAdapter.Item()
                                        .setFile(item.getFile())
                                        .setName(item.getName())
                                        .setKeywordStartIndex(item.getMatchStartIndex())
                                        .setKeywordLength(item.getMatchLength())
                                        .setCount(item.getCount())
                                        .setThumbList(item.getThumbList())
                                )
                                .toList()
                );
    }
}

File: app/src/main/java/cn/intret/app/picgo/screens/adapter/brvah/FolderListAdapterUtils.kt
package cn.intret.app.picgo.screens.adapter.brvah

import cn.intret.app.picgo.model.image.FolderModel
import cn.intret.app.picgo.model.image.ImageFolder
import cn.intret.app.picgo.screens.adapter.SectionedFolderListAdapter
import cn.intret.app.picgo.utils.whenNotNullNorEmpty
import com.annimon.stream.Stream
import java.util.*

object FolderListAdapterUtils {

    fun folderModelToSectionedFolderListAdapter(model: FolderModel): SectionedFolderListAdapter {
        val sections = LinkedList<SectionedFolderListAdapter.Section>()

        val containerFolders = model.containerFolders
        var i = 0
        val s = containerFolders!!.size
        while (i < s) {
            sections.add(parentFolderToItem(containerFolders[i]))
            i++
        }

        return SectionedFolderListAdapter(sections).setHighlightItemName(model.isT9FilterMode())
    }

    private fun parentFolderToItem(containerFolder: FolderModel.ContainerFolder): SectionedFolderListAdapter.Section {
        val section = SectionedFolderListAdapter.Section()

        return section.setName(containerFolder.mName)
                .setFile(containerFolder.file)
                .setItems(
                        Stream.of(containerFolder.folders)
                                .map { item ->
                                    SectionedFolderListAdapter.Item()
                                            .setFile(item.file)
                                            .setName(item.name)
                                            .setKeywordStartIndex(item.getMatchStartIndex())
                                            .setKeywordLength(item.getMatchLength())
                                            .setCount(item.count)
                                            .setThumbList(item.thumbList)
                                }
                                .toList()
                )
    }

    fun imageFolderToFolderListAdapterItem(imageFolder: ImageFolder): FolderListAdapter.Item {
        return FolderListAdapter.Item()
                .setName(imageFolder.name)
                .setCount(imageFolder.count.toLong())
                .setDirectory(imageFolder.file)
                .setThumbList(imageFolder.thumbList)
    }

    fun imageFolderToExpandableFolderItem(folder: ImageFolder): FolderItem {
        return FolderItem()
                .apply {
                    file = folder.file
                    count = folder.count
                    setNameAndCreateSearchUnit(folder.name)
                    setTitle(folder.name)
                    setThumbList(folder.thumbList)
                    setMatchKeywords(folder.matchKeywords?.toString())
                    setMatchLength(folder.getMatchLength())
                    setMatchStartIndex(folder.getMatchStartIndex())
                    setMediaFiles(folder.getMediaFiles())
                    setPinyinSearchUnit(folder.getPinyinSearchUnit())
                }
    }

    fun containerFolderToFolderSection(containerFolder: FolderModel.ContainerFolder): FolderSection {
        return FolderSection().apply {
            setTitle(containerFolder.mName)
            file = containerFolder.file

            val folders = containerFolder.folders
            val folderItems = Stream.of(folders)
                    .map { imageFolderToExpandableFolderItem(it) }
                    .toList()

            subItems = folderItems
        }
    }

    fun folderModelToExpandableFolderAdapter(model: FolderModel): ExpandableFolderAdapter {
        val items = model.containerFolders.whenNotNullNorEmpty { it.map { containerFolderToFolderSection(it) } }
        return ExpandableFolderAdapter(items)
    }
}
